aggregator/list.go

```diff
@@ -182,39 +185,48 @@ func (l *metricList) PushBack(value interface{}) (*list.Element, error) {
 // Close closes the list.
 func (l *metricList) Close() {
 	l.Lock()
-	defer l.Unlock()
-
 	if l.closed {
+		l.Unlock()
 		return
 	}
 	l.closed = true
-}
-
-func (l *metricList) Flush() {
-	flushStart := l.nowFn()
+	l.Unlock()
 
-	// NB(xichen): it is important to determine ticking start time within the time lock
-	// because this ensures all the actions before `start` have completed if those actions
-	// are protected by the same read lock.
-	l.timeLock.Lock()
-	consumeStart := l.nowFn()
-	resolution := l.resolution
-	l.timeLock.Unlock()
-	alignedStartNanos := consumeStart.Truncate(resolution).UnixNano()
-	l.flushBefore(alignedStartNanos, consumeType)
+	// Unregister the list outside the list lock to avoid holding the list lock
+	// while attempting to acquire the flush manager lock potentially causing a
+	// deadlock.
+	if err := l.flushMgr.Unregister(l); err != nil {
+		l.log.Errorf("error unregistering list: %v", err)
+	}
+}
 
-	flushDuration := l.nowFn().Sub(flushStart)
-	l.metrics.flushDuration.Record(flushDuration)
+func (l *metricList) Flush(cutoverNanos, cutoffNanos int64) {
+	start := l.nowFn()
+	alignedNowNanos := l.alignedNowNanos()
+
+	// NB(xichen): If alignedNowNanos <= cutoverNanos, the aggregated metrics should
+	// be discarded because they are aggregated before traffic was cut over. On the
+	// other hand, if alignedNowNanos > cutoffNanos, the aggregated metrics should
+	// also be discarded because they are aggregated after traffic was cut off.
+	if alignedNowNanos > cutoverNanos && alignedNowNanos <= cutoffNanos {
```

Comment:
super nit: perhaps switch the case statements to match the comment? Currently the comment discusses the conditions for dropping metrics but the if statements checks if they can be consumed.

Reply:
Yeah I prefer to handle the common case first, will leave it as is unless you feel strongly.

Reply:
Fine by me. Definitely a super nit :)